### PR TITLE
new Transports for AspnetCore 2.1

### LINKED ISSUE #250: Refactoring Transport once more (aspnetcore 2.1)
Hi @chkr1011  im planning to adapt the Connection abstraction of aspnetcore 2.1 for mqttnet. these abstractions provide a unified way to handle tcp, tls and websockets.

common:
https://github.com/aspnet/KestrelHttpServer/blob/dev/src/Connections.Abstractions/ConnectionContext.cs

tcp:
https://github.com/aspnet/KestrelHttpServer/blob/dev/src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs

websocket:
https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.Http.Connections/Internal/Transports/WebSocketsTransport.cs

I see two ways of integration. What would you prefere?

Option 1: Implement IMqttServerAdapter, IMqttChannelAdapter for tcp in MQTTnet.AspnetCore (that depends on aspnetcore already just need to increment version to 2.1 once released)

Option 2: remove IMqttChannel, IMqttChannelAdapter, ... and let MQTT.Net depend on Kestrel 

what I expect from the change:

1. perf boost
2. less errors

### PR DESCRIPTION
**Pull request by JanEggers**
fixes #250 

the new transports are completely optional meaning aspnetcore project still works with 2.0 and the core api  is  mostly unchanged. 

if a user of this package wants to use the new server they have to upgrade kestrel to 2.1 to get the new api.

if a user wants to use the new api on client side they also have to depend on the aspnetcore package because we dont have a seperate packages for client and server.

the scope of the new transports may be extended later on to provide websocket and tls abilities but tls is currently not available should be ready in aspnet 2.2, with that version there is also an announcement for tcp clients so maybe we can remove that code again. 

perf wise we get roughly 10x improvement in throughput and only use 1/3 memory:

old api: 

|                Method |     Mean |    Error |   StdDev | Rank |      Gen 0 | Allocated |
|  -------------------- |---------:|---------:|---------:|-----:|-----------:|----------:|
|   Send_10000_Messages | 438.6 ms | 7.701 ms | 6.827 ms |    1 | 20812.5000 |  83.52 MB |

new api:

|              Method |     Mean |     Error |    StdDev |     Gen 0 | Allocated |
|-------------------- |---------:|----------:|----------:|----------:|----------:|
| Send_10000_Messages | 49.32 ms | 0.2145 ms | 0.2006 ms | 6375.0000 |  25.73 MB |

### CODE REVIEW

**Comment by JanEggers** [comment]
@chkr1011 can you please drop a comment if you cnsider accepting this. and if not close it i have done 3 rebases already and will do a final one if you intend to merge

**Comment by chkr1011** [comment]
Oh sorry I forgot to complete the review. In general this is all OK for me. I am also playing around with AspNetCore 2.1 at the moment. I added one important comment for me (the event handling).

File: Tests/MQTTnet.TestApp.AspNetCore2/MQTTnet.TestApp.AspNetCore2.csproj [hunk 1]
```diff
@@ -1,7 +1,7 @@
-<Project Sdk="Microsoft.NET.Sdk.Web">
+﻿<Project Sdk="Microsoft.NET.Sdk.Web">
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp2.0</TargetFramework>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
     <TypeScriptToolsVersion>Latest</TypeScriptToolsVersion>
```
[the remark this answers is not in the document]
Comment (JanEggers) [reply Tests/MQTTnet.TestApp.AspNetCore2/MQTTnet.TestApp.AspNetCore2.csproj]: you can use the old api with 2.0, but  the new api needs 2.1

File: Frameworks/MQTTnet.AspnetCore/MqttConnectionContext.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+        public async Task<MqttBasePacket> ReceivePacketAsync(TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            var input = Connection.Transport.Input;
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                ReadResult readResult;
+                ReadingPacketStarted?.Invoke(this, EventArgs.Empty);
+
```
[the remark this answers is not in the document]
Comment (JanEggers) [reply Frameworks/MQTTnet.AspnetCore/MqttConnectionContext.cs]: then i have to move it after the read async call

**Comment by JanEggers** [comment]
can you please have a look again, while testing i also found a bug see:
https://github.com/chkr1011/MQTTnet/pull/298/commits/410bc211a546f7d2e424b45270c0e13c2825ff22#diff-d570014deaf9906f44d19a1bddb188c0R22

**Comment by JanEggers** [comment]
what was the breaking change u fixed in 66048931a2bae650582e638b010cee0104a797f8

**Comment by chkr1011** [comment]
It was nothing special. More likely a merge because I added a parameter for the packet reader to remove overhead from the benchmark. Your implementation was overall good (as usual) 😄

**Pull request merged**